File: ChatGPT/Intermediarias/validate_bst/validate_bst.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        inorder = []

        def traverse(node):
            if not node:
                return True
            if not traverse(node.left):
                return False
            if inorder and node.val <= inorder[-1]:
                return False
            inorder.append(node.val)
            return traverse(node.right)

        return traverse(root)

# Exemplo de execução
if __name__ == "__main__":
    # Criação de uma árvore binária
    #       2
    #      / \
    #     1   3
    node1 = TreeNode(1)
    node3 = TreeNode(3)
    root = TreeNode(2, node1, node3)

    # Cria uma instância da solução e verifica se a árvore é uma BST
    sol = Solution()
    result = sol.isValidBST(root)
    print("A árvore é uma BST válida:", result)

    # Criação de uma árvore binária que não é uma BST
    #       5
    #      / \
    #     1   4
    #        / \
    #       3   6
    node1 = TreeNode(1)
    node3 = TreeNode(3)
    node6 = TreeNode(6)
    node4 = TreeNode(4, node3, node6)
    root = TreeNode(5, node1, node4)

    # Verifica se a segunda árvore é uma BST
    result = sol.isValidBST(root)
    print("A segunda árvore é uma BST válida:", result)

File: ChatGPT/Intermediarias/validate_bst/validate_bst.java
class TreeNode {

    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int x) {
        val = x;
    }
}

public class Solution {

    private TreeNode prev = null;

    public boolean isValidBST(TreeNode root) {
        return inorder(root);
    }

    private boolean inorder(TreeNode node) {
        if (node == null) {
            return true;
        }

        // Visita a subárvore esquerda
        if (!inorder(node.left)) {
            return false;
        }

        // Verifica se o nó atual é maior que o anterior visitado
        if (prev != null && node.val <= prev.val) {
            return false;
        }

        // Atualiza o nó anterior para o atual
        prev = node;

        // Visita a subárvore direita
        return inorder(node.right);
    }

    public static void main(String[] args) {
        // Exemplo de árvore que é uma BST
        TreeNode root = new TreeNode(2);
        root.left = new TreeNode(1);
        root.right = new TreeNode(3);

        Solution sol = new Solution();
        System.out.println(
            "A árvore é uma BST válida: " + sol.isValidBST(root)
        );

        // Exemplo de árvore que não é uma BST
        TreeNode root2 = new TreeNode(5);
        root2.left = new TreeNode(1);
        root2.right = new TreeNode(4);
        root2.right.left = new TreeNode(3);
        root2.right.right = new TreeNode(6);

        System.out.println(
            "A segunda árvore é uma BST válida: " + sol.isValidBST(root2)
        );
    }
}
